**출처 : 파이썬 머신러닝 판다스 데이터 분석, 오승환 지음, 정보문화사**

### 데이터 사전 처리

- [**누락 데이터 처리**](#누락-데이터-처리)
    - [**누락 데이터 확인**](#누락-데이터-확인)
    - [**누락 데이터 제거**](#누락-데이터-제거)
    - [**누락 데이터 치환**](#누락-데이터-치환)    

- [**중복 데이터 처리**](#중복-데이터-처리)
    - [**중복 데이터 확인**](#중복-데이터-확인)
    - [**중복 데이터 제거**](#중복-데이터-제거)

- [**데이터 표준화**](#데이터-표준화)
    - [**단위 환산**](#단위-환산)
    - [**자료형 변환**](#자료형-변환)

- [**범주형(카테고리) 데이터 처리**](#범주형(카테고리)-데이터-처리)
    - [**구간 분할**](#구간-분할)
    - [**더미 변수**](#더미-변수)

- [**정규화**](#정규화)

- [**시계열 데이터**](#시계열-데이터)
    - [**다른 자료형을 시계열 객체로 변환**](#다른-자료형을-시계열-객체로-변환)
        - [**문자열을 Timestamp로 변환**](#문자열을-Timestamp로-변환)    
        - [**Timestamp를 Period로 변환**](#Timestamp를-Period로-변환)            
    - [**시계열 데이터 만들기**](#시계열-데이터-만들기)
        - [**Timestamp 배열**](#Timestamp-배열)
        - [**Period 배열**](#Period-배열)
    - [**시계열 데이터 활용**](#시계열-데이터-활용)
        - [**날짜 데이터 분리**](#날짜-데이터-분리)
        - [**날짜 인덱스 활용**](#날짜-인덱스-활용)

----

# 누락 데이터 처리

* 사전 처리(Preprocessing)
    - 머신러닝 등 데이터 분석에서 정확도는 분석 데이터의 품질에 좌우됨.
    - 데이터 품질을 높이기 위해서 누락 데이터, 중복 데이터 등 오류를 수정하고 분석 목적에 맞게 변형이 필요
    - 수집한 데이터를 분석하기 적합하도록 사전 처리가 필요
    
* 유효한 데이터 값이 존재하지 않는 누락 데이터를 NaN(Not a Number)로 표시

* 머신러닝 분석 모형에 데이터를 입력하기 전에 누락 데이터를 제거하거나 다른 적절한 값으로 대체하는 과정 필요
    - 누락 데이터가 많아지면 데이터의 품질이 떨어지고, 머신러닝 분석 알고리즘의 왜곡 현상이 발생

## 누락 데이터 확인

* 누락 데이터 확인 예제
    - Seaborn 라이브러리의 'titanic' 데이터셋 사용
        - 'deck'열에 NaN 값이 존재. 이 승객의 경우 몇 번 데크에 승선했는지 데이터가 없다는 뜻

In [2]:
# -*- coding: utf-8 -*-

# 라이브러리 불러오기
import seaborn as sns

# titanic 데이터셋 가져오기
df = sns.load_dataset('titanic')
df

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

* info() 메소드로 데이터프레임의 요약 정보 출력
    - 각 열에 속하는 데이터 중에서 유효한(non-null, 즉 NaN값이 아닌) 값의 개수를 보여줌.
    - RangeIndex를 보면 각 열에 891개의 데이터가 존재
    - 'deck' 열에는 203개의 유효한 범주형 데이터가 존재
    - 따라서 'deck' 열에는 누락데이터가 688개(=891-203) 존재

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    object  
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    object  
 8   class        891 non-null    category
 9   who          891 non-null    object  
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    object  
 13  alive        891 non-null    object  
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), object(5)
memory usage: 80.6+ KB


* value_counts() 메소드를 이용하여 'deck' 열에 688개의 누락 데이터가 있는 것을 파악할 수도 있음
    - 누락 데이터의 개수를 확인하기 위해서 dropna=False 옵션 지정
    - 그렇지 않으면 NaN 값을 제외하고 유효한 데이터의 개수만을 구함.  

In [7]:
# deck 열의 NaN 개수 계산하기
nan_deck = df['deck'].value_counts(dropna=False)
nan_deck

NaN    688
C       59
B       47
D       33
E       32
A       15
F       13
G        4
Name: deck, dtype: int64

* 누락 데이터를 직접적으로 찾는 방법 isnull() 메소드와 notnull() 메소드가 있음.
    - isnull() : 누락 데이터면 True 반환, 유효한 데이터가 존재하면 False 반환
    - notnull() : 유효한 데이터가 존재하면 True 반환, 누락 데이터면 False 반환

In [8]:
# isnull() 메서드로 누락 데이터 찾기
df.head().isnull()

survived  pclass    sex    age  sibsp  parch   fare  embarked  class  \
0     False   False  False  False  False  False  False     False  False   
1     False   False  False  False  False  False  False     False  False   
2     False   False  False  False  False  False  False     False  False   
3     False   False  False  False  False  False  False     False  False   
4     False   False  False  False  False  False  False     False  False   

     who  adult_male   deck  embark_town  alive  alone  
0  False       False   True        False  False  False  
1  False       False  False        False  False  False  
2  False       False   True        False  False  False  
3  False       False  False        False  False  False  
4  False       False   True        False  False  False

In [9]:
# notnull() 메서드로 누락 데이터 찾기
df.head().notnull()

survived  pclass   sex   age  sibsp  parch  fare  embarked  class   who  \
0      True    True  True  True   True   True  True      True   True  True   
1      True    True  True  True   True   True  True      True   True  True   
2      True    True  True  True   True   True  True      True   True  True   
3      True    True  True  True   True   True  True      True   True  True   
4      True    True  True  True   True   True  True      True   True  True   

   adult_male   deck  embark_town  alive  alone  
0        True  False         True   True   True  
1        True   True         True   True   True  
2        True  False         True   True   True  
3        True   True         True   True   True  
4        True  False         True   True   True

In [10]:
# isnull() 메서드로 누락 데이터 개수 구하기
df.head().isnull().sum(axis=0) # 참의 합을 구함.

survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       0
class          0
who            0
adult_male     0
deck           3
embark_town    0
alive          0
alone          0
dtype: int64

## 누락 데이터 제거

* 누락 데이터가 들어 있는 열 또는 행을 삭제하는 방법
    - 열을 삭제하면 분석 대상이 갖는 특성(변수)이 제거
    - 행을 삭제하면 분석 대상의 관측값(레코드)가 제거
    
* 예 : 'titanic' 데이터셋의 각 열(변수)에 누락 데이터가 몇 개씩 포함되어 있는지 확인
    - isnull() 메소드와 value_counts() 메소드를 사용
        - 'age'열에 177개, 'embarked'열에 2개, 'deck'열에 688개, 'embark_town'열에 2개      

In [11]:
# -*- coding: utf-8 -*-

# 라이브러리 불러오기
import seaborn as sns

# titanic 데이터셋 가져오기
df = sns.load_dataset('titanic')
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [12]:
missing_df = df.isnull()
missing_df.head()

survived  pclass    sex    age  sibsp  parch   fare  embarked  class  \
0     False   False  False  False  False  False  False     False  False   
1     False   False  False  False  False  False  False     False  False   
2     False   False  False  False  False  False  False     False  False   
3     False   False  False  False  False  False  False     False  False   
4     False   False  False  False  False  False  False     False  False   

     who  adult_male   deck  embark_town  alive  alone  
0  False       False   True        False  False  False  
1  False       False  False        False  False  False  
2  False       False   True        False  False  False  
3  False       False  False        False  False  False  
4  False       False   True        False  False  False

In [14]:
missing_df.columns

Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town',
       'alive', 'alone'],
      dtype='object')

In [18]:
missing_df['deck'].value_counts()

True     688
False    203
Name: deck, dtype: int64

In [19]:
type(missing_df['deck'].value_counts())

pandas.core.series.Series

In [20]:
missing_df['deck'].value_counts()[True]

688

* 전체 891명의 승객 중에서 688명의 데크('deck' 열) 데이터가 누락
    - 누락 데이터가 차지하는 비율이 높음
    - 'deck' 열의 누락 데이터를 삭제하여 분석에서 제외
    - dropna() 메소드에 thresh=500 옵션을 적용하여, NaN 값을 500개 이상 갖는 모든 열을 삭제
        - 'deck' 열만 이 조건에 해당되어 제거

* 전체 891명의 승객 중에서 177명은 나이(age)에 대한 데이터가 없음.
    - 승객의 나이가 데이터 분석의 중요한 변수라면, 나이 데이터가 없는 승객의 레코드(행)을 제거하는 것이 좋음.
    - dropna() 메소드에 subset을 'age'열로 한정하면, 'age'열의 행 중에서 NaN값이 모든 행(axis=0)을 삭제
        - 기본값으로 how='any' 옵션이 적용되는데, NaN 값이 하나라도 존재하면 삭제한다는 뜻.
        - how='all' 옵션으로 입력하면 모든 데이터가 NaN값일 경우에만 삭제가 됨.

* 891개의 행 중에서 나이 데이터가 누락된 177개 행을 삭제하고 나머지 714개의 행을 df_age에 저장

## 누락 데이터 치환

* 데이터셋의 품질을 높일 목적으로 누락 데이터를 무작정 삭제하면, 어렵게 수집한 데이터를 활용하지 못할 수도 있음.
    - 데이터 분석의 정확도는 데이터의 품질 외에도 제공되는 데이터의 양에도 상당한 영향을 받음.
    - 데이터 중에서 일부가 누락되어 있더라도 나머지 데이터를 최대한 살려서 데이터 분석에 활용.
    
* 누락 데이터를 평균값, 중간값, 최빈값 등으로 대체
    - fillna() 메소드로 편리하게 처리
        - fillna() 메소드는 새로운 객체를 반환하기에 원본 객체를 변경하려면 inplace=True 옵션을 추가


* 평균(mean)으로 누락 데이터를 대체하는 방법
    - 앞의 예에서 승객의 나이 데이터가 누락된 행을 제거하지 않고,
    - 대신 'age'열의 나머지 승객의 평균나이로 치환
    - 'age'열에 들어 있는 값들의 평균을 계산하여 mean_age에 저장
    - mean() 메소드를 적용하면 NaN을 제외하고 평균을 계산함.(중간값은 median() 사용)
    - fillna() 메소드에 mean_age를 인자로 전달하면 NaN을 찾아서 mean_age 값으로 치환

* 다음 예제에서 'age' 열의 유효 데이터의 평균(29.699118)으로 대체   

* 가장 많이 나타나는 값으로 NaN을 대체하기
    - 승선도시를 나타내는 'embark_town' 열에 있는 NaN을 
      승객들이 가장 많이 승선한 도시의 이름을 찾아서 NaN을 치환
        - value_counts() 메소드를 사용하여 승선도시별 승객 수를 찾고
        - idxmax() 메소드로 가장 큰 값을 갖는 도시를 찾음.
        - 829행의 NaN값을 포함해서 누락데이터들은 Southampton으로 변경

* 누락 데이터가 NaN으로 표시되지 않은 경우
    - 데이터셋 중에는 누락 데이터가 NaN으로 입력되지 않은 경우도 많음.
        - 예 : 숫자 0이나 문자 '-', '?' 등과 같은 값으로 입력
    - 판다스에서 누락 데이터를 다루려면 replace() 메소드를 활용하여 
      Numpy에서 지원하는 np.nan으로 변경해 주는 것이 좋음.
    - 단, np.nan을 사용하기 위해서는 'import numpy as np'와 같이 Numpy 라이브러리를 먼저 임포트.
    - '?'을 np.nan으로 치환 예 : df.replace('?', np.nan, inplace=True)

* 이웃하고 있는 데이터로 NaN를 대체하기
    - fillna() 메소드
        - method='ffill' 옵션 : NaN이 있는 행의 직전 행에 있는 값으로 대체
        - method='bfill' 옵션 : NaN이 있는 행의 바로 다음 행에 있는 값으로 대체

* 다음 예는 'ffill' 옵션을 사용하여 829행의 NaN 값을 바로 앞에 위치한 828행의 Queenstown으로 변경

----

# 중복 데이터 처리

* 데이터프레임에서 각 행은 분석 대상이 갖고 있는 모든 속성(변수)에 대한 관측값(레코드)을 뜻함.

* 하나의 데이터셋에서 동일한 관측값이 2개 이상 중복되는 경우 중복 데이터를 찾아서 삭제해야 함.
    - 동일한 대상이 중복으로 존재하는 것이므로 분석 결과를 왜곡시킴

## 중복 데이터 확인

* 동일한 관측값의 중복 여부 확인은 duplicated() 메소드 이용
    - 전에 나온 행들과 비교하여 중복되는 행이면 True를 반환하고, 처음 나오는 행에 대해서는 False를 반환
    - 데이터프레임에 duplicated() 메소들를 적용하면 각 행의 중복 여부를 나타내는 불린형 시리즈를 반환
    - 0행과 1행이 동일할 경우
        - 0행은 처음 나오는 값이므로 False 판정, 1행은 앞의 0행과 중복되기 때문에 True

* 데이터프레임의 열은 시리즈 객체이므로, duplicated() 메소드를 적용할 수 있음.

### 중복 데이터 제거

* 중복 데이터 제거하는 drop_duplicates() 메소드
    - 중복되는 행을 제거하고 고유한 관측값을 가진 행들만 남김.
    - 원본 객체를 변경하려면 inplace=True 옵션을 추가

* 다음 예제에서 1행의 데이터는 앞에 이웃하고 있는 0행의 데이터와 중복되므로 제거됨.

* drop_duplicates() 메소드의 subset 옵션에 '열 이름의 리스트'를 전달
    - subset 옵션에 해당하는 열을 기준으로 중복 여부 판별
    - 예 : df의 'c2', 'c3' 열을 기준으로 판별하면 0행과 1행, 3행과 4행의 데이터가 각각 중복됨.

----

# 데이터 표준화

* 다양한 형태의 데이터셋
    - 여러 곳에서 수집한 자료들은 단위 선택, 대소문자 구분, 약칭 활용 등 여러 가지 원인에 의해 다양한 형태로 표현
    - 서로 다른 단위가 섞여 있거나 같은 대상을 다른 형식으로 표현한 경우가 의외로 많음.
    - 동일한 대상을 표현하는 방법에 차이가 있으면, 분석의 정확도는 현저히 낮아짐.

* 데이터 포맷을 일관성 있게 표준화하는 작업이 필요   

## 단위 환산

* 단위 환산 필요성
    - 같은 데이터셋 안에서 서로 다른 측정 단위를 사용하면, 전체 데이터의 일관성 측면(연산 수행 등)에서 문제 발생
    - 외국 데이터의 도량형 단위(영미권의 마일, 야드, 온스 등)를 한국에서 사용하는 미터, 평, 그램 등으로 변환

* 예제 : UCI 자동차 연비 데이터셋
    - 'mpg' 열은 영미권에서 사용하는 '갤런당 마일(mile per gallon)' 단위로 연비를 표시
    - 한국에서 익숙한 표기법인 '리터당 킬로미터(km/L)' 단위로 변환
        - 1마일은 1.60934km이고, 1갤런은 3.78541리터. 따라서 1mpg는 0.425km/L
    - 0행에 들어 있는 차량의 연비는 18mpg
    - 한국식 연비 표현으로 변환한 'kpl' 열의 0행에 해당하는 값은 리터 당 7.65킬로미터.
    - round(2)은 소수점 아래 둘째자리 반올림을 뜻함.
        
    

## 자료형 변환

* 숫자가 문자열(object)로 저장된 경우에 숫자형(int 또는 float)으로 변환
    - dtypes 속성을 사용하여 데이터프레임을 구성하는 각 열의 자료형 확인
        - dtypes 속성 대신 info() 메소드 사용해도 각 열의 자료형을 확인할 수도 있음.
    - 'horsepower' 열은 숫자형이 적절
    - 출시연도를 나타내는 'model year'열과 출시국가를 뜻하는 'origin'열은 카테고리를 나타내는 범주형이 적절

* 'horsepower' 열은 문자열을 뜻하는 object 자료형
    - 엔진 출력의 크기를 나타내는 데이터인 만큼 숫자형으로 변환하는 것이 적절
        - CSV 파일의 '?' 때문에 데이터프레임으로 변환 과정에서 문자열로 인식된 것   

* 'horsepower' 열의 문자열 '?'를 NaN 값으로 변환
    - dropna(axis=0) 메소드로 NaN 값이 들어 있는 모든 행을 삭제
    - 이제 'horsepower' 열에는 숫자형으로 변환 가능한 값들만 남는다.


* astype('float')를 이용하여 문자열을 실수형으로 변환
    - 실수형 대신 정수형으로 변환하려면 'float' 대신 'int'를 입력
    - dtypes 속성을 사용하여 변환된 결과가 맞는지 확인

* 'origin' 열에는 정수형 데이터 1, 2, 3이 들어 있지만, 실제로는 국가 이름인 'USA', 'EU', 'JPN'을 뜻함.
    - replace() 메소드를 사용하여 각 숫자 데이터를 국가이름으로 바꿔주면 문자열을 나타내는 object 자료형을 자동 변경

* 'origin'열의 국가이름은 문자열 데이터(object)임.
    - 값을 확인해 보면 3개의 국가이름이 계속 반복됨.
    - 유한 개의 고유값이 반복적으로 나타나는 경우에는 범주형(category) 데이터로 표현하는 것이 효율적
    - astype('category') 메소드를 이용하면 범주형 데이터로 변환
    - 범주형을 다시 문자열로 변환하려면 astype('str') 메소드를 사용

* 'model year' 열
    - sample() 메소드로 'model year' 열에서 무작위로 3개의 행을 선택해서 출력해보기

* 모델 출시연도가 숫자로 기록, 자료형은 정수형을 나타내는 int64
    - 연도를 뜻하기 때문에 숫자형도 괜찮지만,
    - 연도는 시간적인 순서가 의미가 있음.
    - 데이터는 숫자 형태를 갖더라도 범주형으로 표현하는 것이 적절

----

# 범주형(카테고리) 데이터 처리

## 구간 분할

* 연속 데이터를 일정한 구간(bin)으로 나누어서 분석
    - 가격, 비용, 효율 등 연속적인 값을 일정한 수준이나 정도를 나타내는 이산적인 값을 나타내어 구간별 차이를 드러냄

* 구간 분할(binning)
    - 연속 변수를 일정한 구간으로 나누고, 각 구간을 범주형 이산 변수로 변환하는 과정
    - 판다스 cut() 함수를 이용하면 연속 데이터를 여러 구간으로 나누고 범주형 데이터로 변환할 수 있음.
    
* 구간 분할 예
    - 엔진 출력을 나타내는 'horsepower에서, 엔진 출력을 숫자로 표시하는 대신,
    - '저출력', '보통출력', '고출력' 등 구간으로 나누어서 표시할 경우
    - 경계값을 구하는 방법 중에서 Numpy 라이브러리의 histogram() 함수를 활용
        - 나누려는 구간(bin) 개수를 bins 옵션에 입력하면 각 구간에 속하는 값의 개수(count)와 경계값 리스트(bin_dividers)를 반환
    - 아래 예제에서는 모두 4개의 경계값을 생성하고 3개의 구간이 만들어진다 (46\~107.3 구간, 107.3\~168.6 구간, 168.6\~230 구간)
    

* 판다스 cut() 함수의 옵션을 설정
    - 앞에서 구한 경계값의 리스트(bin_dividers)를 bins 옵션에 할당하고 각 구간의 이름(bin_names)을 labels 옵션에 할당
    
* 각 구간의 낮은경계값을 포함하기 위해 include_lowest=True 옵션 사용
    - 'horsepower' 열의 숫자 데이터를 3개의 구간에 할당하고,
    - 각 구간의 이름을 '저출력', '보통출력', '고출력' 으로 입력하여 'hp_bin'열에 저장

## 더미 변수

* 'horsepower' 열의 숫자형 연속 데이터를 'hp_bin' 열의 범주형 데이터로 변환하였다.

* 범주형 데이터를 회귀분석 등 머신러닝 알고리즘에 바로 사용할 수 없는 경우가 있음
    - 컴퓨터가 인식 가능한 입력값으로 변환이 필요
    - 이럴 때 숫자 0 또는 1로 표현되는 더미 변수(dummy variable)을 사용
        - 여기서 0과 1은 수의 크고 작음을 나타내지 않고, 어떤 특성(feature)이 있는지 없는지 여부만을 표시
        - 해당 특성이 존재하면 1로 표현, 존재하지 않으면 0으로 구분하는 개념
        
* 원핫인코딩(one-hot encoding)
    - 범주형 데이터를 컴퓨터가 인식할 수 있도록 숫자 0과 1로 구성된 원핫벡터(one-hot vector)로 변환
    
* 판다스 get_dummies() 함수를 사용하면, 범주형 변수의 모든 고유값을 각각 새로운 더미 변수로 변환
    - 'hp_bin' 열의 고유값 3개가 각각 새로운 더미 변수 열의 이름이 됨.
    - 각 더미 변수가 본래 속해 있던 행에는 1이 입력되고, 속하지 않았던 다른 행에는 0이 입력됨.

* sklearn 라이브러리를 이용한 원핫인코딩
    - 데이터프레임 df의 'hp_bin' 열에 들어 있는 범주형 데이터를 0, 1을 원소로 갖는 원핫벡터로 변환
        - 결과는 선형대수학에서 정의하는 희소행렬(sparse matrix)로 정리됨.

* 예제에서는 1차원 벡터를 2차원 행렬로 변환하고 다시 희소행렬로 변환
    - 희소행렬은 (행, 열) 좌표와 값 형태로 정리됨.
    - 앞에서 (0, 1)은 0행의 1열 위치를 말하고, 데이터 값은 숫자 1.0이 입력됨.

----

# 정규화

* 정규화의 필요성
    - 각 변수(데이터프레임의 열)에 들어 있는 숫자 데이터의 상대적 크기 차이 때문에 머신러닝 분석 결과가 달라질 수 있음.
        - 예: A 변수는 0\~1000 범위의 값을 가지고, B 변수는 0\~1 범위의 값을 가질 경우, 상대적으로 큰 숫자 값을 갖는 A 변수의 영향이 더 커짐.
    - 숫자 데이터의 상대적인 크기 차이를 제거할 필요가 있음.


* 정규화(normalization)
    - 데이터의 상대적 크기에 대한 영향을 줄이기 위해 데이터 범위를 변환
    - 각 열(변수)에 속하는 데이터 값을 동일한 크기 기준으로 나눈 비율로 나타내는 것
    - 정규화를 거친 데이터의 범위는 0\~1 또는 -1\~1이 됨

* 정규화 방법
    - 각 열(변수)의 데이터 중에서 최대값(max)과 최소값(min)을 뺀 값으로 나누는 방법
        - 각 열 데이터에서 해당 열의 최소값을 뺀 값을 분자로 하고, 해당 열의 최대값과 최소값의 차를 분모로 하는 수를 계산
        - 정규화 후 값의 범위 0 \~ 1
        - X값의 정규화 수식 : (X - min) / (max - min)
    - 예 : 'horsepower'열의 최대값은 230이고 최소값은 46, 이들의 차인 184를 이용하여 정규화하면, 최소 0부터 최대 1사이의 범위로 변환


----

 # 시계열 데이터

* 시계열 데이터
    - 일정 기간에 대해 시간의 함수로 표현되는 데이터.
    - 예: 과거 5년간의 연도별 매상액, 3시간마다의 태풍 위치 등
    - 주로 예측 업무에 사용

* 판다스의 시간 표시 방식
    - Timestamp : 특정한 시점을 기록
    - Period : 두 시점 사이의 일정한 기간

## 다른 자료형을 시계열 객체로 변환

### 문자열을 Timestamp로 변환

* 판다스 to_datetime() 함수
    - 문자열 등 다른 자료형을 판다스 Timestamp를 나태내는 datetime64 자료형으로 변환
    
* 예
    - 주식 시장에서 거래되는 A 종목의 거래 데이터를 정리한 CSV 파일을 read_csv() 함수를 이용하여 불러옴.
    - head() 메소드로 데이터프레임 살펴보면,
    - 'Date' 열에 날짜 데이터가 들어 있는데, info() 메소드로 확인하면 문자열(object)임  

* 'Date' 열의 날짜 데이터를 판다스 Timestamp 객체로 변경
    - 'Date' 열을 to_datetime() 함수의 인자로 전달하면 문자열(object) 데이터를 datetime64 자료형으로 변환


* 변환된 데이터를 'new_Date' 열에 담아서 데이터프레임 df에 추가

* info() 메소드로 'new_Date' 열에 들어 있는 데이터 값들이 datetime64 자료형임을 확인

* 'new_date' 열의 개별 원소 데이터를 type() 함수로 확인하면 Timestamp 객체라는 것을 확인

* 'new_Date' 열을 데이터프레임 df의 행 인덱스로 설정하고, 'Date' 열을 제거
    - 이렇게 시계열 값을 행 인덱스로 지정하면 판다스는 DatetimeIndex로 저장함.
    - 시계열 인덱스 클래스를 지원하기 때문에 시간 순서에 맞춰 인덱싱 또는 슬라이싱 하기가 편리함.

### Timestamp를 Period로 변환

* 판다스 to_period() 함수
    - Timestamp 객체를 일정한 기간을 나타내는 Period 객체로 변환


* to_period() 함수 예제
    - 3개의 날짜 데이터를 Timestamp로 변환하고, 기간 옵션을 달리 하여 Period 객체를 지정
    - DatetimeIndex가 PeriodIndex로 변환되고, 자료형은 datetime64에서 period로 변환
    - freq 옵션
        - 'D' : 1일의 기간을 나타냄
        - 'M' : 1개월의 기간을 나타냄
        - 'A' : 1년의 기간을 나타냄 (1년이 끝나는 12월이 기준)

* freq 옵션의 종류
    - https://pandas.pydata.org/pandas-docs/stable/user_guide/timeseries.html

## 시계열 데이터 만들기

### Timestamp 배열

* 판다스 date_range() 함수
    - 여러 개의 날짜(Timestamp)가 들어 있는 배열 형태의 시계열 데이터 생성
    - range() 함수로 숫자 배열을 만드는 것과 비슷
    
* 예제 : '2019-01-01'부터 한 달 간격으로 각 달의 시작 날짜를 6개 생성
    - 날짜 범위의 시작점으로 '2019-01-01'을 설정, 끝을 따로 정하지 않음(end=None)
    - periods=6은 Timestamp를 6개 생성한다는 뜻
    - freq='MS'에서 'M'은 월을 뜻하고 'S'는 월의 시작일(1일)을 나타냄.
    - tz='Asia/Seoul'은 한국 시간대를 설정   

* 시간 간격을 다르게 설정
    - freq='M'으로 입력하면 월의 마지막 날짜를 생성
    - freq='3M' : 3개월 간격의 마지막 날짜

### Period 배열

* 판다스 period_range() 함수
    - 여러 개의 기간(Period)이 들어 있는 시계열 데이터를 생성
    
* 예제 : '2019-01', '2019-02', '2019-03'을 원소로 Period 배열
    - 날짜 범위의 시작점으로 '2019-01-01'을 설정, 날짜 범위의 끝을 따로 정하지 않음(end=None)
    - period=3 옵션은 Period를 3개 만든다는 뜻
    - freq='M'에서 'M'은 월을 나타냄  

* 기간을 다르게 설정
    - freq='H' : 1시간 간격
    - freq='2H' : 2시간 간격

## 시계열 데이터 활용

### 날짜 데이터 분리

* 연-월-일 날짜 데이터에서 일부를 분리하여 추출

* 연-월-일 정보를 연, 월, 일 각각으로 구분하는 방법
    - 날짜 정보가 있는 'Date' 열을 to_datetime() 함수에 전달하면 Timestamp로 변환됨.
    - 변환된 결과를 'new_Date' 열에 담아서 데이터프레임에 추가
    - dt 속성을 이용하여 'new_Date' 열의 연-월-일 정보에서  연, 월, 일을 개별적으로 추출
    - dt.year로 추출한 연도는 'Year' 열에, dt.month로 추출한 월 데이터는 'Month' 열에, dt.day로 추출한 일 데이터는 'Day' 열에 저장  

* Timestamp 객체를 Period 객체로 변환하는 to_period() 메소드를 적용하여, 연-월-일 중에서 연-월 또는 연도를 추출
    - 'Date_yr' 열에는 연도를 나타내는 값이 저장
    - 'Date_m' 열에는 연-월을 나타내는 값이 저장

* 추출한 날짜 정보를 데이터프레임의 행 인덱스로 지정할 수도 있음.
    - 예제 : 'Date_m' 열을 데이터프레임의 새로운 행 인덱스로 지정

### 날짜 인덱스 활용

* Timestamp/Period로 구성된 열을 행 인덱스로 지정
    - Timestamp로 구성된 열을 행 인덱스로 지정하면 DatetimeIndex라는 고유 속성으로 변환
    - Period로 구성된 열을 행 인덱스로 지정하면 PeriodIndex라는 고유 속성으로 변환
    - 날짜 인덱스를 활용하면 시계열 데이터에 대한 인덱싱과 슬라이싱이 편리함
    

* 예제
    - to_datetime() 메소드를 이용하여 'Date' 열 데이터를 Timestamp로 변환하고 'new_Date'열에 저장
    - set_index() 메소드로 'new_Date' 열을 데이터프레임의 행 인덱스로 지정
        - 행 인덱스 값들의 자료형은 datetime64[ns]

* 날짜 인덱스 사용의 장점
    - 연-월-일 중에서 내가 필요로 하는 레벨을 선택적을 인덱싱할 수 있음.
    - 연도만을 기준으로 하거나 연-월 또는 연-월-일 기준으로 선택 가능
    - 예: df['2018']과 같이 추출하려는 날짜는 입력하면 해당 날짜에 해당하는 모든 행을 선택할 수 있음.

* Timestamp 객체로 표시된 두 날짜 사이의 시간 간격을 구할 수 있음
    - 예제 : 어떤 날짜로부터 경과한 날짜를 계산하여 행 인덱스로 지정하고, 데이터를 선택하는 과정을 살펴보자